### PR TITLE
Update for Universal Flag Configuration

### PR DESCRIPTION
**Pull request by aarsilv**
_Eppo Internal_:
🎟️ **Ticket:** [FF-2003 - Update Android SDK for UFC](https://linear.app/eppo/issue/FF-2003/update-android-sdk-for-ufc)
🗺️ **UFC Design Review:** [UFC: Universal Flag Configuration](https://www.notion.so/eppo/UFC-Universal-Flag-Configuration-f08ac43f6e6542e59a0416e6b738d8f0?pvs=4)

This pull request updates our Android SDK to be able to ingest and use the new Universal Flag Configuration format, and test that against our UFC test cases.

The high-level changes are:
- Updated DTOs to store the new structure
- Updated adapters for serializing and deserializing the new structure
- Strict typing only (carried out by a modified EppoValue)
- Utilities and support for full obfuscation of the configuration
- Updated and new tests for all of the above

### CODE REVIEW

File: eppo/src/test/java/cloud/eppo/android/UtilsTest.java [hunk 1]
```diff
@@ -10,15 +21,31 @@
...
+    }
+
     @Test
     public void testGetISODate() {
         String isoDate = Utils.getISODate(new Date());
         assertNotNull("ISO date should not be null", isoDate);
 
         // Verify the format
-        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US);
```
Comment (aarsilv) [inline eppo/src/test/java/cloud/eppo/android/UtilsTest.java:48]: Our configs have milliseconds

File: Makefile [hunk 2]
```diff
@@ -35,9 +37,9 @@ test-data:
 	rm -rf $(testDataDir)
 	mkdir -p $(tempDir)
 	git clone -b ${branchName} --depth 1 --single-branch ${githubRepoLink} ${gitDataDir}
-	cp ${gitDataDir}/rac-experiments-v3.json ${testDataDir}
-	cp ${gitDataDir}/rac-experiments-v3-hashed-keys.json ${testDataDir}
-	cp -r ${gitDataDir}/assignment-v2 ${testDataDir}
+	cp ${gitDataDir}/ufc/flags-v1.json ${testDataDir}
+	cp ${gitDataDir}/ufc/flags-v1-obfuscated.json ${testDataDir}
+	cp -r ${gitDataDir}/ufc/tests ${testDataDir}
```
Comment (aarsilv) [inline Makefile:42]: Use UFC test data

File: eppo/src/androidTest/java/cloud/eppo/android/helpers/TestCaseValue.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package cloud.eppo.android.helpers;
+
+import com.google.gson.JsonElement;
+
+import java.util.List;
+
+import cloud.eppo.android.dto.EppoValue;
+
+public class TestCaseValue extends EppoValue {
```
[the remark this answers is not in the document]
Comment (schmit) [reply eppo/src/androidTest/java/cloud/eppo/android/helpers/TestCaseValue.java:9]: Should we update the test cases?

File: eppo/src/main/java/cloud/eppo/android/EppoClient.java [hunk 4]
```diff
@@ -82,219 +113,179 @@ private void refreshConfiguration(InitializationCallback callback) {
...
+        boolean typeMatch;
+        switch (expectedType) {
+            case BOOLEAN:
+                typeMatch = value.isBoolean();
+                break;
+            case INTEGER:
+                typeMatch = value.isNumeric()
+                        // Java has no isInteger check so we check value == floor == ceil
+                        && Math.floor(value.doubleValue()) == Math.ceil(value.doubleValue())
+                        && Math.floor(value.doubleValue()) == value.doubleValue();
```
Comment (schmit) [inline eppo/src/main/java/cloud/eppo/android/EppoClient.java:189]: Easier to check that `x mod 1 = 0` instead?

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                Variation decodedVariation = new Variation();
+                decodedVariation.setKey(base64Decode(variation.getKey()));
+                EppoValue decodedValue = EppoValue.nullValue();
+                if (!variation.getValue().isNull()) {
+                    String stringValue = base64Decode(variation.getValue().stringValue());
+                    switch (flag.getVariationType()) {
+                        case BOOLEAN:
+                            decodedValue = EppoValue.valueOf("true".equals(stringValue));
+                            break;
+                        case INTEGER:
```
Comment (schmit) [inline eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:108]: Doesn't Java have an integer type? https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        if (split.getShards() == null || split.getShards().isEmpty()) {
+            // Default to matching if no explicit shards
+            return true;
+        }
+
+        boolean allShardsMatch = true;
+        for (Shard shard : split.getShards()) {
+            if (!matchesShard(shard, subjectKey, totalShards, isObfuscated)) {
+                allShardsMatch = false;
+                break;
```
Comment (schmit) [inline eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:138]:
```suggestion
                return false;
```

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            }
+        }
+        return allShardsMatch;
```
Comment (schmit) [inline eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:141]:
```suggestion
        return true;
```
do not need `allShardsMatch` when you return early in the for loop above

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        String hashKey = salt+"-"+subjectKey;
+        int assignedShard = Utils.getShard(hashKey, totalShards);
+        boolean inRange = false;
+        for (Range range : shard.getRanges()) {
+            if (assignedShard >= range.getStart() && assignedShard < range.getEnd()) {
+                inRange = true;
+                break;
+            }
+        }
+        return inRange;
```
Comment (schmit) [inline eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:158]: same:
```suggestion
        for (Range range : shard.getRanges()) {
            if (assignedShard >= range.getStart() && assignedShard < range.getEnd()) {
                return true;
            }
        }
        return false;
```

**Review by schmit — APPROVED** [review]
🚀

File: eppo/src/androidTest/java/cloud/eppo/android/helpers/TestCaseValue.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aarsilv) [reply eppo/src/androidTest/java/cloud/eppo/android/helpers/TestCaseValue.java:9]: No, because its for the `assignment` field and we are expecting a JSON assignment from getJsonAssignment, so I think it all makes sense.

File: eppo/src/main/java/cloud/eppo/android/EppoClient.java (hunk 4 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/EppoClient.java:189]: Ha yes, that would be easier 😅

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java (hunk 5 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:108]: It does! But `EppoValue` does not distinguish and just loads the JavaScript number, which does not have an integer type.

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java (hunk 7 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:141]: Yeah not sure why I did it this way, I think I didn't have the empty check in the guard clause before so was being clever to handle the empty shards case but now we can just return

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java (hunk 8 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:158]: This we need as-is as I think if there is no range, then we consider it always in range. Or I could do this with a guard clause up top. Allthough all other things equal, I have a slight bias towards fewer `return` statements.

The other thing with shards was probably mirroring this structure.

File: eppo/src/main/java/cloud/eppo/android/ConfigurationStore.java [hunk 9]
```diff
@@ -65,63 +66,64 @@ public void loadFromCache(CacheLoadCallback callback) {
                     return;
                 }
 
-                flags = configResponse.getFlags();
-                Log.d(TAG, "Cache loaded successfully");
+                flags = new ConcurrentHashMap<>(configResponse.getFlags());
```
Comment (leoromanovsky) [inline eppo/src/main/java/cloud/eppo/android/ConfigurationStore.java:69]: kudos for using `ConcurrentMap`

File: eppo/src/main/java/cloud/eppo/android/RuleEvaluator.java [hunk 10]
```diff
@@ -59,8 +87,12 @@ private static boolean evaluateCondition(SubjectAttributes subjectAttributes, Ta
             Version valueSemVer = null;
             Version conditionSemVer = null;
             try {
-                valueSemVer =  Version.parse(value.stringValue());
-                conditionSemVer = Version.parse(condition.getValue().stringValue());
+                valueSemVer = Version.parse(attributeValue.stringValue());
+                String conditionSemVerString = condition.getValue().stringValue();
+                if (isObfuscated) {
```
Comment (leoromanovsky) [inline eppo/src/main/java/cloud/eppo/android/RuleEvaluator.java:92]: 👍

File: eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java [hunk 11]
```diff
@@ -36,27 +43,39 @@ public void onCacheLoadSuccess() {
 
             @Override
             public void onCacheLoadFail() {
-                // no-op; fall-back to Fetch
+                cacheLoadInProgress.set(false);
+                Log.d(TAG, "Did not initialize from cache");
+                // If cache loading failed, and fetching failed, fire off the failure callback if not yet done so
```
Comment (leoromanovsky) [inline eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java:48]: 👍

File: eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java [hunk 12]
```diff
@@ -7,27 +7,34 @@
...
     public ConfigurationRequestor(ConfigurationStore configurationStore, EppoHttpClient client) {
         this.configurationStore = configurationStore;
         this.client = client;
     }
 
     public void load(InitializationCallback callback) {
+        // We have two at-bats to load the configuration: loading from cache and fetching
+        // The below variables help them keep track of each other's progress
+        AtomicBoolean cacheLoadInProgress = new AtomicBoolean(true);
+        AtomicReference<String> fetchErrorMessage = new AtomicReference<>(null);
```
Comment (leoromanovsky) [inline eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java:30]: Can you see this behavior being configurable based on if the user wants to wait for the network request to finish?

File: eppo/src/test/java/cloud/eppo/android/FlagConfigResponseDeserializerTest.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+public class FlagConfigResponseDeserializerTest {
+
+    private final Gson gson = new GsonBuilder()
+            .registerTypeAdapter(FlagConfigResponse.class, new FlagConfigResponseAdapter())
+            .create();
+
+    @Test
+    public void testDeserializePlainText() throws IOException {
+        File testRac = new File("src/androidTest/assets/flags-v1.json");
```
Comment (sameerank) [inline eppo/src/test/java/cloud/eppo/android/FlagConfigResponseDeserializerTest.java:40]: This is coming from https://github.com/Eppo-exp/sdk-test-data/blob/main/ufc/flags-v1.json right?

So this would actually be

```suggestion
        File testUfc = new File("src/androidTest/assets/flags-v1.json");
```

?

File: eppo/src/main/java/cloud/eppo/android/dto/Range.java [hunk 14]
```diff
@@ -1,6 +1,6 @@
 package cloud.eppo.android.dto;
 
-public class ShardRange {
+public class Range {
```
Comment (sameerank) [inline eppo/src/main/java/cloud/eppo/android/dto/Range.java:3]: Why was this renamed? It's still the returned type by `createShardRange` so it's still a shard range? Should that function be renamed too for consistency or does `Range` served other purposes as well?

File: eppo/src/main/java/cloud/eppo/android/dto/Allocation.java [hunk 15]
```diff
@@ -1,25 +1,68 @@
...
+
+    private boolean doLog;
+
+    public String getKey() {
+        return key;
+    }
+
+    public void setKey(String key) {
+        this.key = key;
+    }
```
Comment (sameerank) [inline eppo/src/main/java/cloud/eppo/android/dto/Allocation.java:27]: Are constructor parameters not supported in this version of Java? Lots of getters and setters 😮

File: eppo/src/main/java/cloud/eppo/android/dto/EppoValue.java [hunk 16]
```diff
@@ -1,113 +1,110 @@
...
+        switch(this.type) {
+            case BOOLEAN:
+                return this.boolValue.toString();
+            case NUMBER:
+                return this.doubleValue.toString();
+            case STRING:
+                return this.stringValue;
+            case ARRAY_OF_STRING:
+                return String.join(" ,", this.stringArrayValue);
+            default: // NULL
```
Comment (sameerank) [inline eppo/src/main/java/cloud/eppo/android/dto/EppoValue.java:106]: Or just be explicit about it

```suggestion
            case NULL:
```

File: eppo/src/main/java/cloud/eppo/android/util/Utils.java [hunk 17]
```diff
@@ -48,10 +82,33 @@ public static void validateNotEmptyOrNull(String input, String errorMessage) {
...
+    public static Date parseUtcISODateElement(JsonElement isoDateStringElement) {
+        if (isoDateStringElement == null || isoDateStringElement.isJsonNull()) {
+            return null;
+        }
+        String isoDateString = isoDateStringElement.getAsString();
+        Date result = null;
+        try {
+            result = isoUtcDateFormat.parse(isoDateString);
+        } catch (ParseException e) {
+            /* no-op; leave result null for now */
```
Comment (sameerank) [inline eppo/src/main/java/cloud/eppo/android/util/Utils.java:94]: If this catch scenario expected to happen during normal operation? Or would it be a symptom of something being broken/used incorrectly?

**Review by sameerank — APPROVED** [review]
Great work! Also lots of great feedback here already!

File: eppo/src/main/java/cloud/eppo/android/dto/adapters/DateAdapter.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.Date;
+
+/**
+ * This adapter for Date allows gson to serialize to UTC ISO 8601 (vs. its default of local timezone)
+ */
+public class DateAdapter implements JsonSerializer<Date> {
+
+    @Override
+    public JsonElement serialize(Date src, Type typeOfT, JsonSerializationContext context) {
+        return new JsonPrimitive(getISODate(src));
```
[the remark this answers is not in the document]
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/dto/adapters/DateAdapter.java:20]: Happy to do this if we want to drop MS, was thinking just keeping things consistent with how we'll receive it in the UFC.

File: eppo/src/main/java/cloud/eppo/android/ConfigurationStore.java (hunk 9 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/ConfigurationStore.java:69]: It was there before I modified things so I can't take credit. I did make the higher-level abstractions just using `Map` and `ConcurrentHashMap` as the concrete implementation.

File: eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java (hunk 12 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/ConfigurationRequestor.java:30]: Yes! I think everything should confirm to [this plan](https://www.notion.so/eppo/Client-SDK-Caching-and-Fetching-Behavior-a9c8caf98c7e4a27b3d4c7751ad65cd9?pvs=4#4d1575abf9e64f0cbd25cf68df470d0d).

For now, this is mirroring how it behaved before (minus the shared preferences, which got ripped out).

File: eppo/src/test/java/cloud/eppo/android/FlagConfigResponseDeserializerTest.java (hunk 13 above)
Comment (aarsilv) [reply eppo/src/test/java/cloud/eppo/android/FlagConfigResponseDeserializerTest.java:40]: Yes! I should rename the local variable

File: eppo/src/main/java/cloud/eppo/android/dto/EppoValue.java (hunk 16 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/dto/EppoValue.java:106]: ha true, will do

File: eppo/src/main/java/cloud/eppo/android/util/Utils.java (hunk 17 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/util/Utils.java:94]: This would be expected if parsing a base-64 encoded (obfuscated) date. I can expand the comment to indicate it's expected.

File: eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java (hunk 8 above)
Comment (aarsilv) [reply eppo/src/main/java/cloud/eppo/android/FlagEvaluator.java:158]: Correction: for this, if there is no range, it's not in range, so changing per your suggestion, and it will have the same structure as the other change you suggested, which I like the parallelism.
